Get the provisioning ISO working

amstewart opened the pull request:
This patchset...
1. switches the default provisioning scheme for the automated VM creation scripts to `efi-ab` - which is the correct scheme for RAUC targets. This change is necessary to get the [provisioning ptest](https://github.com/ni/meta-nilrt/blob/nilrt/master%2Fdunfell/recipes-core/initrdscripts/files/ptest/test_ni_provisioning.sh) working. 
2. deprecates the non-QEMU virtual machine flavors from the buildVM script. They have not been officially supported for a long time and they do not support the `efi-ab` boot scheme. There will be a concomitant PR to the AZDO components to fixup the script arguments in the makefiles.
3. adds a new `test_provisioning` script which uses QEMU and tcl-expect to instrument the `init-restore-mode-ptest` test. The AZDO PR will add the new subcomponent which makes use of this script to block the AZDO build when the ISO has created a nonfunctional VM.

## Testing

* Ran the new provisioning test many times both manually and through nibuild.

@ni/rtos @jpautler

amstewart committed:
buildVM: switch provisioning scheme to efi-ab

The efi-ab provisioning scheme is required to properly configure EFI on
the RAUC-enabled QEMU VMs. Set it as the default to facilitate automated
testing of the recovery media.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
buildVM: deprecate non-QEMU VM targets

We have not supported non-QEMU NILRT VMs in a long time. Remove these
machine targets to simplify the codebase. Users who need these machines
will need to provision their own, using the recovery media ISO.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
tests: add new provisioning test

Add a new, QEMU-based test for the provisioning ISO. The test will use
tcl-expect to drive a test session on an already-provisioned VM. It will
install the init-restore-mode-ptest package from either the default opkg
feeds, or alternate feeds, and run its ptest.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
tests: deprecate old provisioning test

Now that we have a new provisioning test for the RAUC provisioning ISO,
deprecate the old, migration-IPK-based test. The old test is no longer
relevant and is named similarly enough that it would be otherwise
confusing to keep around.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

shruthi-ravi commented on scripts/buildVM-files/ni_provisioning.answers:
```diff
@@ -3,8 +3,6 @@
 # Full automated installation; yes on all options + poweroff on finish
 PROVISION_UNSUPPORTED_TARGET="y"
 PROVISION_REPARTITION_TARGET="y"
-PROVISION_RESTORE_MINIMAL_IMAGE="y"
```
What's this change about?

amstewart replied:
As far as I can tell, `PROVISION_RESTORE_MINIMAL_IMAGE` is not used by any of our scripts and - I would assume - refers to the nxg "minimal" images which we have deprecated. I removed the setting for cleanliness.

shruthi-ravi commented on scripts/tests/provisioning/test.expect:
```diff
@@ -0,0 +1,65 @@
+# shutdown
+expect {
+	-re "$USER@.*# $" { send -- "halt\r" }
+	timeout { exit 1 }
```
Can you add some text to describe what timed out? The ptest-runner hung?

amstewart replied:
Sure; will fix.

shruthi-ravi commented on scripts/tests/provisioning/test_provisioning.sh:
```diff
@@ -0,0 +1,163 @@
+
+usage() {
+	local exit_code=${1:-2}
+	test $exit_code -eq 0 && exec 1>&2
```
Shouldn't this be -neq? Redirect stdout to stderr when exit_code neq 0. Have I misunderstood?

amstewart replied:
Yep. Good catch. Will fix.

shruthi-ravi reviewed (approved):
Minor. Nice work!

jpautler commented on scripts/buildVM-files/ni_provisioning.answers:
```diff
@@ -3,8 +3,6 @@
 # Full automated installation; yes on all options + poweroff on finish
 PROVISION_UNSUPPORTED_TARGET="y"
```
I assume you removed PROVISION_RESTORE_MINIMAL_IMAGE because it is never referenced by the provisioning scripts (was removed at some point in the past maybe?). If so, can you remove PROVISION_UNSUPPORTED_TARGET as well since it seems to never be referenced either.

amstewart replied:
Correct. I'll see if I can remove the UNSUPPORTED_TARGET var as well.

shruthi-ravi replied:
Can you update the commit message to reference these changes while you're at it? :)

jpautler commented on scripts/tests/provisioning/test.expect:
```diff
@@ -0,0 +1,65 @@
+	timeout { exit 1 }
+}
+
+expect EOF
```
I'm not sure, but should we have a timeout associated with the run-vm script exiting within a reasonable amount of time? For example, if the target VM gets stuck during shutdown should we consider that an error?

This test will probably become a reference for future tests, so if we think that would be useful maybe we want to illustrate it here? If not, feel free to drop this suggestion.

amstewart replied:
The shutdown command will follow the default timeout for the expect script, which is set at 90 seconds. If it takes longer than that, I will print an `TIMEOUT` message and the entire expect process (and VM) will exit with code 1. That will trigger the `set -e` in the `test_provisioning` script and propagate the error to the subcomponent. ie. A hang during shutdown should cause the test to fail sanely.

jpautler replied:
I see. I think I was thrown off by the fact that the expect call just above this had an explicit timeout case. Does that explicit timeout case do anything different than if you had just left it off and relied on the default timeout behavior?

amstewart replied:
See my latest patch-rev.

jpautler replied:
If I understand the behavior properly, I think the **expect eof** still needs a timeout clause if we want to return an error code and print a message.

amstewart committed:
tests/provisioning: make TIMEOUT more verbose

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart committed:
fixup! tests: add new provisioning test

The usage() function should print to stderr *except* when the user
specifically asks for help.

amstewart committed:
buildVM: remove deprecated options from answers file

Several answers file options are no longer supported and should be
removed from the answers file for clarity.

Refusing to bump the answers file version because there is nothing about
the new file which breaks compatibility with the old interpreters.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
## Patch V2
* Removed the `PROVISION_UNSUPPORTED_TARGET` option from the answers file, because it is no longer needed.
* Add more version TIMEOUT messages to `test.expect`.
* Fixup the `usage()` output redirection logic in `test_provisioning.sh`.

amstewart committed:
test_provisioning: fixup timeout messages

If a timeout is not explicitly defined, the default action in tcl-expect
is to NOOP and continue with the test. That's not what we want, so
define a timeout action for all stages in the test.

Also fixup the TIMEOUT messages by adding a carriage return character.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

jpautler commented on scripts/tests/provisioning/test.expect:
```diff
@@ -34,20 +34,22 @@ expect {
+		send "opkg install init-restore-mode-ptest\r"
+	}
+	timeout {
+		send_user "TIMEOUT waiting for opkg to configure and ptests to install.\r"
```
Isn't this expect clause waiting for login to complete so it can **send** the opkg commands rather than waiting **for** the opkg commands to complete? I think this same issue applies to each timeout clause below.

amstewart replied:
No. The `timeout` clause applies to the time it takes to execute all actions within the parent `expect` clause, not the time between this clause and the previous.

When I artificially cause the `configure opkg` clause to timeout, for example, you get this:
```bash
...
NI-Unknown-52542d724e76 login: root

TIMEOUT waiting for opkg to configure and ptests to install.
INFO: Feed server (2320) terminated.
removed '/tmp/test_provisioning.nDsHD/feed.0'
removed '/tmp/test_provisioning.nDsHD/feed.1'
removed '/tmp/test_provisioning.nDsHD/feed.2'
/home/usr0/projects/ni-central/src/rtos/rtosimagesd/rtos_nilinuxrt/source/nilrt/scripts/tests/provisioning/test_provisioning.sh: line 110:  2320 Terminated              python3 -m http.server $IPK_FEED_PORT  (wd: /tmp/test_provisioning.nDsHD)
make[2]: *** [objects/test_provisioning/linuxU/x64/gcc-4.7-oe/release/ni-linux-rt-x64-qemu/provisioning.log] Error 1

```

amstewart replied:
^ This is wrong. See Patch V5 below.

jpautler commented on scripts/tests/provisioning/test.expect:
```diff
@@ -0,0 +1,65 @@
+expect {
+	"login: " { send "root\r" }
+	timeout {
+		send_user "TIMEOUT while waiting for getty."
```
Should this line get a \r as well?

amstewart replied:
Yeah. Also, in the process of triple checking my understanding to your comment above, I have figured out a better way to write these `send_user` calls. Will fix in V4.

amstewart committed:
fixup! test_provisioning: fixup timeout messages

Using the newline character, instead of the carriage return, is better
for `send_user` in a GNU make context. When using \n before and after
the message, MAKE will avoid its tendancy to drop the first character of
the message output and also to overwrite the message line with the next
stderr output.

Prevously, messages would sometimes look like this:
```
IMEOUT waiting for opkg to configure and ptests to install.
INFO: Feed server (2473) terminated.
```

or this
```
INFO: Feed server (2473) terminated. and ptests to install.
```

Now they are printed in their completion on thier own line.

amstewart commented:
## Patch V4

Using the newline character, instead of the carriage return, is better
for `send_user` in a GNU make context. When using \n before and after
the message, MAKE will avoid its tendancy to drop the first character of
the message output and also to overwrite the message line with the next
stderr output.

Prevously, messages would sometimes look like this:
```
IMEOUT waiting for opkg to configure and ptests to install.
INFO: Feed server (2473) terminated.
```

or this
```
INFO: Feed server (2473) terminated. and ptests to install.
```

Now they are printed in their completion on their own line.

amstewart committed:
test_provisioning: fixup TIMEOUT messages

The `timeout` cases in `expect` clauses apply to a timeout waiting for
the parent clause to trigger - not while waiting for the parent to
complete. Move the timeout messages to their correct locations.

Signed-off-by: Alex Stewart <alex.stewart@ni.com>

amstewart commented:
## Patch V5
* @jpautler was correct about the timeout locations. My original verification was incorrect and the messages should have been printing information about why the *previous* clause failed. I moved the messages into their correct locations.
* Used the `$USER` expect variable for getty login, as I had originally intended.

### Testing

* Used an intentionally-wrong username for getty login to confirm that the timeout messages are speaking to the correct failures.
* Re-verified that the test still passes with these changes (when using the correct getty username).

shruthi-ravi commented on scripts/tests/provisioning/test.expect:
```diff
@@ -25,7 +25,7 @@ spawn [lindex $argv 0] -s -- -device e1000,netdev=n1 -netdev user,id=n1,ipv6=off
-	"login: " { send "root\r" }
+	"login: " { send "$USER\r" }
 	timeout {
 		send_user "\nTIMEOUT while waiting for getty.\n"
```
> the messages should have been printing information about why the previous clause failed...

If that's the case, what does this timeout apply to?

shruthi-ravi replied:
Nvm, I misunderstood your description.

shruthi-ravi commented on scripts/tests/provisioning/test.expect:
```diff
@@ -47,7 +47,7 @@ expect {
 	}
 	timeout {
-		send_user "\nTIMEOUT waiting for opkg to configure and ptests to install.\n"
+		send_user "\nTIMEOUT waiting for shell.\n"
```
And do you need a 90s timeout for the shell to appear?

amstewart replied:
You shouldn't *need* one. But that's how I have it written, yes.
